cupboard_app/test_api.py

```diff
@@ -1480,3 +1481,118 @@ def test_create_user_api_without_email_in_token(self, mock_decode):
                 'message': UserViewSet.MISSING_USER_INFO
             }
         )
+
+
+class CustomIngredientsApi(TestCase):
+    unit1 = None
+    unit2 = None
+
+    def setUp(self):
+        """
+        Sets up a test database with test values
+        """
+        self.user1 = User.objects.create(
+            username=USER_VALID_TOKEN_PAYLOAD.get('sub'),
+            email=USER_VALID_TOKEN_PAYLOAD.get(CUPBOARD_EMAIL_CLAIM)
+        )
+
+        self.ing1 = CustomIngredient.objects.create(
+            user=self.user1,
+            name='test_ingredient1',
+            type='test_type1'
+        )
+        self.ing2 = CustomIngredient.objects.create(
+            user=self.user1,
+            name='test_ingredient2',
+            type='test_type1'
+        )
+
+    @patch.object(TokenBackend, 'decode')
+    def test_create_custom_ingredient(self, mock_decode):
+        """
+        Testing create_custom_ingredient creates a custom ingredient
+        in the database
+        """
+        mock_decode.return_value = USER_VALID_TOKEN_PAYLOAD
+
+        response = self.client.post(
+            reverse(f'{API_VERSION}:add_delete_customingredient'),
+            json.dumps(
+                {
+                    'username': self.user1.username,
+                    'ingredient': 'Beef',
+                    'type': 'Meat'
+                }
+            ),
+            content_type='application/json',
+            HTTP_AUTHORIZATION='Bearer valid-token'
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertDictEqual(
+            response.json(),
+            {'user': self.user1.username, 'name': 'Beef', 'type': 'Meat'}
+        )
+
+    @patch.object(TokenBackend, 'decode')
+    def test_create_custom_ingredient_nonexistant_user(self, mock_decode):
+        """
+        Testing create_custom_ingredient works properly when
+        a username doesn't exist in a custom ingredient
+        """
+        mock_decode.return_value = USER_VALID_TOKEN_PAYLOAD
```

Comment:
It's a good idea to test this, but you would test this by replacing the user in the USER_VALID_TOKEN_PAYLOAD.
```
mock_decode.return_value = {**USER_VALID_TOKEN_PAYLOAD, 'sub': 'Doesn't exist user}
```

Reply:
Changed the line as you mentioned, but I'm getting a CacheKeyWarning that I think is related to my implementation of this. Not sure if you ran into that too.

Reply:
hmmm I'll check it out! I didn't run into this with other things, so there might something else going on.